### ISSUE TITLE
tap: generic body matching

### ISSUE THREAD

**Issue by mattklein123** [envoyproxy/envoy#6107]
Add generic text body matchers that perform matches up to some buffer limit.

**Comment by cpakulski** [comment]
@mattklein123 Since there are no assumptions about the body's structure, I think that only matchers would be:
- contains _token_
- does not contain _token_
where the _token_ is string or regex.

Should the matching process kick in only when content-Type is text/plain, or leave it to a user to configure two matchers?

```
and_match:
  rules:
    - http_request_headers_match:
              headers:
               - name: content-type
                  exact_match: plain/text
    - generic_body_match:
             contains: envoy

```

**Comment by mattklein123** [comment]
@cpakulski I think that sounds about right. I probably wouldn't worry about content type and would allow binary body matching.

**Comment by cpakulski** [comment]
@mattklein123 Thank you. I am going to implement is the way you suggest.

**Comment by cpakulski** [comment]
@mattklein123 I have developed a prototype for generic body matching. As configuration it takes a list of text strings or hex strings. In order to get a match all listed strings must be found in the body. The code is on my branch: https://github.com/cpakulski/envoy/tree/issue/6107.

There are few issues I would like to discuss:
- `onBody` method (https://github.com/cpakulski/envoy/blob/issue/6107/source/extensions/filters/http/tap/tap_config_impl.cc#L177-L215) is called from `HttpPerRequestTapperImpl::onResponseBody` and from `HttpPerRequestTapperImpl::onRequestBody` without indicating the direction. It means that the same instance of BodyMatcher is invoked for request and later on for response. Currently I implemented it using OR (https://github.com/cpakulski/envoy/blob/issue/6107/source/extensions/common/tap/tap_matcher.cc#L162) - when either direction matches the matcher indicates MATCH. There is clear direction distinction for Headers - Request and Response. I am wondering if we want to do the same for Body. There are 2 ways: 1st - use two different types `GenericRequestBodyMatcher` and `GenericResponseBodyMatcher` and 2nd way - add a direction parameter to `GenericBodyMatcher` api (`enum : any, request, response`). There are two other body matchers in the pipeline: JSON and gRPC and the same idea will have to be replicated there.

- For matching I am using `search` method from `buffer` library (https://github.com/cpakulski/envoy/blob/issue/6107/source/extensions/common/tap/tap_matcher.cc#L159). It works but scans entire payload and I see no easy way to limit search to first X bytes. Are you aware of any utility already in the code where I could do text match indicating bytes limit? If not,there are 2 options: develop new text matcher or extend `buffer` library to limit search to indicated number of bytes.
